README
⚠️ **This repository has moved to Codeberg: https://codeberg.org/cdamian/wikicalendar**

This GitHub repository is archived and no longer maintained here.
	Simple calendar extension for mediawiki.
	
	Released under the GPL. Copyright 2005 by Christof Damian
	(christof@damian.net)
	
	I wrote this for http://krass.com/ where I have simple event
	calendar. Examples are at http://krass.com/wiki/Current_events and
	http://krass.com/wiki/Calendar
	
	The idea was to keep the calendar as simple as possible and leave all
	the editing and page creation to mediawiki. It doesn't require any new
	database tables or files.
	
	To use this on your mediawiki site, copy the "calendar" directory into your
	extension directory and put this somewhere at the end of your 
	"LocalSettings.php":
	
	  require_once("$IP/extensions/calendar/Calendar.php");
	
	Then you can use the calendar on any page like this:
	
	<calendar>
	name=mycalendar
	view=year
	</calendar>
	
	The parameters work like this:
	
	"view"
	       can be "days" days from today, "rdays" (reverse days),
	       "week","month","threemonths", "year" (default), "weeks" and "months".
	"name"
	       is a name for the calendar and changes the names of the day pages.
	"format"
	       defaults to '%name_%year_%month_%day' and describes the names
	       of the day pages.
	"formattitle"
	       formats the titles in today/days/week view
	"day","month","year"
	       you can specify the date of the calendar, it defaults to today.
	       currently this doesn't work with "view=week"
	"date"
	       this takes a php strtotime ( http:://www.php.net/strtotime )
	       option and overrides the "day","month","year" option
	       above. examples are:
	
	       date="+1 week"
	       date="next monday"
	       date="+7 days"
	"enddate"
	       just used for the weeks view, this is the last date to display in the
	       weeks
	"weekstart"
	       week starts on 1=monday ... 7=sunday. default is 1
	"skipempty"
	       skip empty days in week view
	"showempty"
	       show inivitation to add entries when a day is empty, default is 1
	"weekformat"
	       format of the week, days, rdays and today view.
	
	       - "text" (default), simple wikitext format
	       - "list" a layout that uses <dl>, <dt> and <dd> (definition
	         list) tags with the 'calendar' class added to the <dl>
	         tag and the calendarToday class to the span tag that
	         contains the weekday title. Merged calendar
	         entries are added as sub lists with the name part of the
	         merged calendar as a link to the page of that
	         calendar.
	       - "table" a layout that uses table tags with also the name of
	         merged calendars as a link to the page of those
	         calendars. The 'calendar' class is added to the table tag and
	         the 'calendarToday' class to the span tag that contains the
	         weekday title.
	"months"
	        number of months for the "months" view
	"mergemonth"
			include the day page in the month, months and year views
	
	format & formattitle accept these substitutions:
	        %name
	        %day,%month,%year
	        any php date() substitution prefixed with '%', like '%m','%d','%Y'
	
	Thanks go to these contributors:
	
	- Amador amador@tiscali.nl (localization, weekformat, showempty)
	- Austin Che (cache fix)
	- Dexter Samida <dextersamida@gmail.com> (today view)
	- Kai Sommer (months view)


ChangeLog
2009-09-20  Christof Damian  <christof@damian.net>

	* fix for multibyte day and month names
	* releasing version 1.16

2009-02-27  Christof Damian  <christof@damian.net>

	* releaseing version 1.15

2008-09-23 Christof Damian  <christof@damian.net>

	* new "mergemonth" option for the month, months and year view

2008-09-16 	Christof Damian  <christof@damian.net>

	* new "weeks" view
	* new "months" view

2006-10-09  Christof Damian  <christof@damian.net>

	* released 1.13

	* fix for mediawiki 1.7

2006-09-29  Christof Damian  <christof@damian.net>

	* released 1.12

	* fix for '[[]]' display when merging and main calendar entry is empty

	* fix for including more than one calendar

	* simple ical format, which I use on krass.com

2006-04-18  Christof Damian  <christof@damian.net>

	* released 1.11

	* 1.10 had a syntax error

	* fix for one date() call

2006-04-06  Christof Damian  <christof@damian.net>

	* released 1.10

	* fixed weekformat=list

	* xml style parameters

	* released 1.9

	* weekformat is in classes now

2006-03-10  Christof Damian  <christof@damian.net>

	* released 1.8

	* "weekformat" option for different layouts

	* "showempty" to supplress "add entry" msgs

	* localized calendar from user preferences

	* fix merge function to pass calendar name

	* The Amador Release

2006-03-01  Christof Damian  <christof@damian.net>

	* released 1.7

	* another fix for the cache (thanks to Austin Che)

	* "skipempty" option

	* "rdays" reverse order of days

2005-11-17  Christof Damian  <christof@damian.net>

	* released 1.6

	* replaced <h3> in week display, to remove TOC in mediawiki 1.5

2005-10-14  Christof Damian  <christof@damian.net>

	* first go at merging calendars

	* add extension credits

2005-10-12  Christof Damian  <christof@damian.net>

	* released 1.5

	* fixed month display again

	* released 1.4

	* included some tests for CalendarClass using PHPUnit2

	* fix weeks not starting on monday

	* fix padding after month ending in end of the week

	* new DaysInMonth() and DayOfWeek() not using the php calendar
	extensions anymore

2005-10-04  Christof Damian  <christof@damian.net>

	* new clearCache() version by Daniel Simon

2005-09-28  Christof Damian  <christof@damian.net>

	* "weekstart" option

2005-09-16  Christof Damian  <christof@damian.net>

	* fix version parsing for rc releases like 1.5rc1

	* changed "<?" to "<?php" in CalendarClass.php, thanks Juanma
	Barranquero

	* new option "formattitle" for the titles in the week/day views
	* format & formattitle both accept php date() parameters now,
	  prefixed with '%'. like this: "%d.%m.%Y".

2005-09-15  Christof Damian  <christof@damian.net>

	* enable double quoted options

	* "date" option, this takes a php strtotime string

	* disable cache for 1.5 (untested)

2005-09-06  Christof Damian  <christof@damian.net>

	* new view="days" with additional parameter "days" for the number
	of days to show. I needed that for krass.com

	* new view="today" option from Dexter Samida <dextersamida@gmail.com>

	I'm not completely happy with it, but it doesn't hurt at the moment

2005-08-29  Christof Damian  <christof@damian.net>

	* fix problem with "format" option




calendar/Calendar.php
<?php

/*

Simple wiki calendar
Copyright (C) 2005 Christof Damian

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

*/

require_once(dirname(__FILE__)."/WikiCalendarClass.php");
require_once(dirname(__FILE__)."/WikiCalendarFormatText.php");
require_once(dirname(__FILE__)."/WikiCalendarFormatList.php");
require_once(dirname(__FILE__)."/WikiCalendarFormatTable.php");
require_once(dirname(__FILE__)."/WikiCalendarFormatIcal.php");


$wgExtensionFunctions[] = "wfCalendarExtension";
$wgExtensionCredits['parserhook'][] = array(
   'name' => 'wikicalendar',
   'author' => 'Christof Damian',
   'url' => 'http://code.google.com/p/wikicalendar/',
   'version' => "1.16",
   'description' => 'simple calendar extension',
   'descriptionmsg' => 'wikicalendar-desc'
   );

function wfCalendarExtension() {
    global $wgParser;
    $wgParser->setHook( "calendar", "renderCalendar" );
}

/**
 * Method to clear MediaWiki Cache on different versions
 *
 * @version 1.1
 * @author Daniel Simon
 */
function clearCache() {
  global $wgVersion;
  if (version_compare($wgVersion,'1.5','>=')) {
    global $wgParser;
    $wgParser->disableCache();
  } elseif (version_compare($wgVersion,'1.4','>=')) {
    global $wgTitle;
    $dbw =& wfGetDB( DB_MASTER );
    $dbw->update( 'cur', array(
			'cur_touched' => $dbw->timestamp( time() + 120 )
		),
        array(
        	'cur_namespace' => $wgTitle->getNamespace(),
            'cur_title' => $wgTitle->getDBkey()
        ), 'CalendarExtension'
	);
  } elseif (version_compare($wgVersion,'1.3','>=')) {
    $wgOut->enableClientCache(false);
  }
}

function renderCalendar( $paramstring, $params = array() , $parser ) {
	global $wgTitle,$wgUser,$wgOut,$wgVersion;

	if (!$parser) {
		$parser = & new Parser();
	}

	$p = array(
    	"view"  => "year",
        "day"   => 0,
        "month" => 0,
        "year"  => 0,
        "days"  => 7,
		"months" => 2,
        "weekstart" => 1,
        "formattitle" => "%j.%n.%Y %l",
        "skipempty" => 0,
        "showempty" => 1,
        "weekformat" => "text",
        "weekdaylen" => 1,
		"enddate" => false,
		"mergemonth" => false
	);

	preg_match_all(
		'/([\w]+)\s*=\s*(?:"([^"]+)"|([^"\s]+))/',
		$paramstring,
		$matches);
	for ($i=0; $i< count($matches[0]); $i++) {
		$p[$matches[1][$i]] = $matches[2][$i].$matches[3][$i];
	}

	foreach (array('view','day','month','year','days','weekstart',
                 'formattitle','format','name','date','skipempty',
                 'showempty','weekformat','weekdaylen','enddate','months',
				'mergemonth'
		) as $i) {
		if (isset($params[$i])) {
			$p[$i] = $params[$i];
		}
	}

	if (!isset($p['name'])) {
		$p['name'] = 'calendar';
	};
	if (!isset($p['format'])) {
		$p['format'] = '%name_%year_%month_%day';
	};

	if (isset($p['date'])) {
		$time = strtotime($p['date']);
		$p['day']   = date('d',$time);
		$p['month'] = date('n',$time);
		$p['year']  = date('Y',$time);
	}

	$cal = new WikiCalendarClass($p["year"],$p["month"],$p["day"]);
	$cal->format = $p['format'];
	$cal->formattitle = $p['formattitle'];
	$cal->name = $p['name'];
	$cal->weekstart = $p['weekstart'];
	$cal->skipempty = $p['skipempty'];
	$cal->showempty = $p['showempty'];
	$cal->weekdaylen = $p['weekdaylen'];
	$cal->mergemonth = $p['mergemonth'];

	if (isset($p["merge"])) {
		$cal->merge = explode(',',$p["merge"]);
	}

	switch ($p['weekformat']) {
		case 'list':
			$cal->weekformat = new WikiCalendarFormatList();
			break;
		case 'table':
			$cal->weekformat = new WikiCalendarFormatTable();
			break;
		case 'ical':
			$cal->weekformat = new WikiCalendarFormatIcal();
			break;
		default:
			$cal->weekformat = new WikiCalendarFormatText();
	}

	switch ($p["view"]) {
		case "week":
			$calstr = $cal->displayDays();
			break;
		case "month":
			$calstr = $cal->displayMonth();
			break;
		case "today":
			$calstr = $cal->displayDays(1);
			break;
		case "threemonths":
			$calstr = $cal->displayThreeMonths();
			break;
		case "months":
			$calstr = $cal->displayMonths($p['months']);
			break;
		case "days":
			$calstr = $cal->displayDays($p["days"]);
			break;
		case "rdays":
			$calstr = $cal->displayDays($p["days"], true);
			break;
		case "weeks":
			$calstr = $cal->displayWeeks($p['enddate']);
			break;
		case "year":
		default:
			$calstr = $cal->displayYear();
	}

	$o = & $parser->parse(
		$calstr,
		$wgTitle,
		ParserOptions::newFromUser($wgUser),
		true,
		false);

	clearCache();

	return $o->getText();
}

?>